pkg/crc/machine/sync.go

```diff
@@ -0,0 +1,119 @@
+package machine
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/code-ready/crc/pkg/crc/logging"
+	"github.com/code-ready/machine/libmachine/state"
+	"golang.org/x/sync/semaphore"
+)
+
+const startCancelTimeout = 15 * time.Second
+
+type Synchronized struct {
+	startLock      *semaphore.Weighted
+	stopDeleteLock *semaphore.Weighted
+	startCancel    context.CancelFunc
+	underlying     Client
+}
+
+func NewSynchronizedMachine(machine Client) *Synchronized {
+	return &Synchronized{
+		startLock:      semaphore.NewWeighted(1),
+		stopDeleteLock: semaphore.NewWeighted(1),
+		underlying:     machine,
+	}
+}
+
+func (s *Synchronized) IsProgressing() bool {
+	if !s.startLock.TryAcquire(1) {
+		return true
+	}
+	defer s.startLock.Release(1)
+	if !s.stopDeleteLock.TryAcquire(1) {
+		return true
+	}
+	defer s.stopDeleteLock.Release(1)
+	return false
+}
+
+func (s *Synchronized) Delete() error {
+	cleanup, err := s.lockForStopOrDelete()
+	defer cleanup()
+	if err != nil {
+		return err
+	}
+	return s.underlying.Delete()
+}
+
+func (s *Synchronized) Start(ctx context.Context, startConfig StartConfig) (*StartResult, error) {
+	if !s.startLock.TryAcquire(1) {
+		return nil, errors.New("cluster is starting")
+	}
+	defer s.startLock.Release(1)
+	ctx, s.startCancel = context.WithCancel(ctx)
+	return s.underlying.Start(ctx, startConfig)
+}
+
+func (s *Synchronized) Stop() (state.State, error) {
+	cleanup, err := s.lockForStopOrDelete()
```

Comment:
I am curious if we shouldn't be more explicit that Delete can only happen AFTER a Stop has been performed. It seems here it is defined as either.

Reply:
Very good question. Today, the CLI doesn't prevent you to delete right after start. 
Delete is a destructive action so, I think, there is no need to wait for the VM to be correctly stopped to delete it.

In the other way, there is no point stopping a VM that is in the process of being deleted. 

`crc delete` is perhaps the wrong word. I see it as a lightweight `crc cleanup`, more like `crc reset`

Reply:
I believe it also do this sometimes, but yeah... I do not think this should prevent the PR.
It is more a detail.

Reply:
> Delete can only happen AFTER a Stop has been performed

@gbraad, I was asking about this some time very early on and I'm pretty sure I was told that both `start` -> `stop` -> `delete` and `start` -> `delete` were valid workflows. As @guillaumerose says, why do we care if VM is destroyed in a nice state or not?  It would seem to me that `stop` only makes sense if the user intends to restart that same VM later. Otherwise `start` -> `delete` seems like a more suitable sequence. Basically just asking out of curiosity...